sway/desktop/render.c

```diff
@@ -70,6 +72,33 @@ static void scissor_output(struct wlr_output *wlr_output,
 	wlr_renderer_scissor(renderer, &box);
 }
 
+static void set_scale_filter(struct wlr_output *wlr_output,
+		struct wlr_texture *texture, enum scale_filter_mode scale_filter) {
+	if (!wlr_texture_is_gles2(texture)) {
+		return;
+	}
+
+	struct wlr_gles2_texture_attribs attribs;
+	wlr_gles2_texture_get_attribs(texture, &attribs);
+
+	switch (scale_filter) {
+	case SCALE_FILTER_LINEAR:
+		glTexParameteri(attribs.target, GL_TEXTURE_MAG_FILTER, GL_LINEAR);
+		break;
+	case SCALE_FILTER_NEAREST:
+		glTexParameteri(attribs.target, GL_TEXTURE_MAG_FILTER, GL_NEAREST);
+		break;
+	case SCALE_FILTER_SMART:
+		if (ceilf(wlr_output->scale) == wlr_output->scale) {
```

Comment:
We need to call `glTexParameteri` with `GL_LINEAR` in the other case to correctly react to live config changes.

I'd suggest setting `scale_filter` to either linear or nearest before the switch if it's set to smart.

Reply:
I'm not sure I understand. You mean when the output scale changes? In my observation that works already, though I'm not actually certain why (or if it's reliable).

Either way, why not set it in an `else` after the integer check from the "smart" case if we want?

Reply:
Yeah, when the scale changes from being an integer to a fraction, we need to set `GL_LINEAR`.

>Either way, why not set it in an else after the integer check from the "smart" case if we want?

Because it duplicates code. Suggested alternative:

```c
if (scale_filter == SCALE_FILTER_SMART) {
	scale_filter = ceilf(wlr_output->scale) == wlr_output->scale ? SCALE_FILTER_NEAREST :
		SCALE_FILTER_LINEAR;
}

switch (scale_filter) {
case SCALE_FILTER_LINEAR:
	glTexParameteri(attribs.target, GL_TEXTURE_MAG_FILTER, GL_LINEAR);
	break;
case SCALE_FILTER_NEAREST:
	glTexParameteri(attribs.target, GL_TEXTURE_MAG_FILTER, GL_NEAREST);
	break;
default:
	assert(false); // unreachable
}
```